File: design-patterns/1-creational/builder/main.java
interface Builder {
    string setOptionA(string a);
    string setOptionB(string b);
    string setOptionC(string c);
}

class ConcreteBuilderA implements Builder {
    private string a;
    private string b;
    private string c;

    @Override
    string setOptionA(string a) {
        this.optionA = a
    }

    @Override
    string setOptionA(string b) {
        this.optionB = b
    }

    @Override
    string setOptionC(string c) {
        this.optionB = c
    }

    public ConcreteResultA getResult() {
        return new ConcreteResultA(a, b, c);
    }
}

class ConcreteBuilderB implements Builder {
    private string a;
    private string b;
    private string c;

    @Override
    string setOptionA(string a) {
        this.optionA = a
    }

    @Override
    string setOptionA(string b) {
        this.optionB = b
    }

    @Override
    string setOptionC(string c) {
        this.optionB = c
    }

    public ConcreteResultB getResult() {
        return new ConcreteResultB(a, b, c);
    }
}

public class ConcreteResultA {
    public string optionA;
    public string optionB;
    public string optionC;

    public ConcreteResultA(string a, string b, string c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }
//     getters, setters
}

public class ConcreteResultB {
    public string optionA;
    public string optionB;
    public string optionC;

    public ConcreteResultB(string a, string b, string c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }

}

// директор
public class Director {
    public void constructProduct(Builder builder) {
        builder.setOptionA("option-1");
        builder.setOptionB("option-2");
        builder.setOptionC("option-3");
    }
}

// client
public class App {
    public static void main(args...) {
        Director d = new Director();

        Builder builderA = new ConcreteBuilderA();
        d.constructProduct(builderA);

        ConcreteResultA result = builderA.getResult();

//      -------------------------
        Builder builderB = new ConcreteBuilderB();
        d.constructProduct(builderB);

        ConcreteResultB result1 = builderB.getResult();
    }
}

File: design-patterns/1-creational/builder/main.go
package main

func main() {
	director := Director{}

	builderA := ConcreteBuilderA{}
	director.ConstructProduct(&builderA)
	productA := builderA.GetResult()

	builderB := ConcreteBuilderB{}
	director.ConstructProduct(&builderB)
	productB := builderB.GetResult()
}

type Builder interface {
	SetOptionA(a string)
	SetOptionB(b string)
	SetOptionC(c string)
}

type ConcreteBuilderA struct {
	A string
	B string
	C string
}

func (aa *ConcreteBuilderA) SetOptionA(a string) {
	aa.A = a
}

func (aa *ConcreteBuilderA) SetOptionB(b string) {
	aa.B = b
}

func (aa *ConcreteBuilderA) SetOptionC(c string) {
	aa.C = c
}

func (aa *ConcreteBuilderA) GetResult() ProductA {
	return ProductA{
		A: aa.A,
		B: aa.B,
		C: aa.C,
	}
}

type ConcreteBuilderB struct {
	A string
	B string
	C string
}

func (bb *ConcreteBuilderB) SetOptionA(a string) {
	bb.A = a
}

func (bb *ConcreteBuilderB) SetOptionB(b string) {
	bb.B = b
}

func (bb *ConcreteBuilderB) SetOptionC(c string) {
	bb.C = c
}

func (bb *ConcreteBuilderB) GetResult() ProductB {
	return ProductB{
		A: bb.A,
		B: bb.B,
		C: bb.C,
	}
}

type ProductA struct {
	A string
	B string
	C string
}

type ProductB struct {
	A string
	B string
	C string
}

type Director struct {
}

func (d *Director) ConstructProduct(b Builder) {
	b.SetOptionA("a")
	b.SetOptionB("b")
	b.SetOptionC("c")
}
